jobbergate-api/jobbergate_api/storage.py

```diff
@@ -11,45 +15,187 @@
+            await engine.dispose()
+        self.engine_map = dict()
+
+    def get_engine(self, override_db_name: str | None = None) -> AsyncEngine:
+        """
+        Get a database engine.
+
+        If the database url is already in the engine map, return the engine stored there. Otherwise, build
+        a new one, store it, and return the new engine.
+        """
+        db_url = build_db_url(
+            override_db_name=override_db_name,
+            force_test=settings.DEPLOY_ENV.lower() == "test",
+        )
+        if db_url not in self.engine_map:
+            self.engine_map[db_url] = create_async_engine(db_url, pool_pre_ping=True)
+        return self.engine_map[db_url]
+
+    def get_session(self, override_db_name: str | None = None) -> AsyncSession:
+        """
+        Get an asynchronous database session.
+
+        Gets a new session from the correct engine in the engine map.
+        """
+        engine = self.get_engine(override_db_name=override_db_name)
+        return AsyncSession(engine)
+
+
+engine_factory = EngineFactory()
+
+
+@dataclass
+class SecureSession:
+    """
+    Provide a container class for an IdentityPayload and AsyncSesson for the current request.
+    """
+
+    identity_payload: IdentityPayload
+    session: AsyncSession
+
+
+def secure_session(*scopes: str, permission_mode: PermissionMode = PermissionMode.ALL):
+    """
+    Provide an injectable for FastAPI that checks permissions and returns a database session for this request.
+
+    This should be used for all secured routes that need access to the database. It will commit the
+    transaction upon completion of the request. If an exception occurs, it will rollback the transaction.
+    If multi-tenancy is enabled, it will retrieve a database session for the database associated with the
+    client_id found in the requesting user's auth token.
+
+    If testing mode is enabled, it will flush the session instead of committing changes to the database.
+
+    Note that the session should NEVER be explicitly committed anywhere else in the source code.
+    """
+
+    async def dependency(
+        identity_payload: IdentityPayload = fastapi.Depends(
+            lockdown_with_identity(*scopes, permission_mode=permission_mode)
+        )
+    ) -> typing.AsyncIterator[SecureSession]:
+        is_test = settings.DEPLOY_ENV.lower() == "test"
+
+        override_db_name = identity_payload.organization_id if settings.MULTI_TENANCY_ENABLED else None
+        session = engine_factory.get_session(override_db_name=override_db_name)
+        nested_transaction = await session.begin_nested()
+        try:
+            yield SecureSession(
+                identity_payload=identity_payload,
+                session=session,
+            )
+            # # In test mode, we should not commit to the database. Instead, just flush to the session
+            # In test mode, we should not commit to the database. Instead, just commit the nested transaction.
```

Comment:
Is it duplicated?

Reply:
Yeah, I was trying it out. It works, so I will clean up the comments and commented out code.

Reply:
Addressed in commit a59185b